src/firewall.py

```diff
@@ -38,87 +35,118 @@ def decode(cls, entry: str) -> "FirewallEntry":
-        """Render the firewall template with the provided allowlist.
-
-        Args:
-            allowlist: The list of FirewallEntry objects to allow.
+    def get_host_ip(self) -> str:
+        """Get the host IP address for the corresponding LXD network.
 
         Returns:
-            str: The rendered firewall template as a string.
+            The host IP address.
         """
-        return self._template.render(allowlist=allowlist, host_ip=self._host_ip)
+        address = execute_command(
+            ["/snap/bin/lxc", "network", "get", self._network, "ipv4.address"]
+        )
+        return str(ipaddress.IPv4Interface(address.strip()).ip)
 
-    def refresh_firewall(self, allowlist: typing.List[FirewallEntry]):
+    def refresh_firewall(self, denylist: typing.List[FirewallEntry]):
         """Refresh the firewall configuration.
 
         Args:
-            allowlist: The list of FirewallEntry objects to allow.
+            denylist: The list of FirewallEntry objects to allow.
         """
+        current_acls = [
+            acl["name"]
+            for acl in yaml.safe_load(
+                execute_command(["lxc", "network", "acl", "list", "-f", "yaml"])
+            )
+        ]
+        if self._ACL_RULESET_NAME not in current_acls:
+            execute_command(["/snap/bin/lxc", "network", "acl", "create", self._ACL_RULESET_NAME])
         execute_command(
-            ["/usr/sbin/nft", "-f", "-"],
-            input=self._render_firewall_template(allowlist).encode("ascii"),
+            [
+                "/snap/bin/lxc",
+                "network",
+                "set",
+                self._network,
+                f"security.acls={self._ACL_RULESET_NAME}",
+            ]
+        )
+        execute_command(
+            [
+                "/snap/bin/lxc",
+                "network",
+                "set",
+                self._network,
+                "security.acls.default.egress.action=allow",
+            ]
+        )
+        acl_config = yaml.safe_load(
+            execute_command(["/snap/bin/lxc", "network", "acl", "show", self._ACL_RULESET_NAME])
+        )
+        host_ip = self.get_host_ip()
+        egress_rules = [
+            {
+                "action": "reject",
+                "destination": host_ip,
+                "destination_port": "1-8079,8081-65535",
+                "protocol": "tcp",
+                "state": "enabled",
+            },
+            {
+                "action": "reject",
+                "destination": host_ip,
+                "protocol": "udp",
+                "state": "enabled",
+            },
+            {
+                "action": "reject",
+                "destination": host_ip,
+                "protocol": "icmp4",
+                "state": "enabled",
+            },
+            {
+                "action": "reject",
+                "destination": "::/0",
+                "state": "enabled",
```

Comment:
What is the effect of this rule please?

Reply:
Reject all IPv6 traffics, since we don't provision a IPv6 network to the runner right now.

Reply:
This feels a little Canonical specific, should that go to the denylist configuration rather than in our code? If someone else installs the charm on an IPv6 network I'm guessing it wouldn't work with this configuration?

Reply:
The LXD network inside the runner has IPv6 disabled as well, so I suspect it will work even with this rule removed.

Reply:
Ah ok